File: 0421-maximum-xor-of-two-numbers-in-an-array/0421-maximum-xor-of-two-numbers-in-an-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
class TrieNode{
    constructor(){
        this.left = null;
        this.right = null;
    }
}
class Trie{
    constructor(){
        this.root = new TrieNode();
    }

    insert(num){
        let curr = this.root;

        for(let i = 31; i >= 0; i--){
            let ith_bit = (num >> i) & 1;

            if(ith_bit == 1){
                if(curr.right == null){
                    curr.right = new TrieNode();
                }
                curr = curr.right;
            }else{
                if(curr.left == null){
                    curr.left = new TrieNode();
                }
                curr = curr.left;
            }
        }
    }
}

function findMaxXor(tr, num){
    let curr = tr.root;
    let maxXor = 0;

    for(let i = 31; i >= 0; i--){
        let ith_bit = (num >> i) & 1;

        if(ith_bit == 0){

            if(curr.right != null){
                maxXor += Math.pow(2, i);
                curr = curr.right;
            }else{
                curr = curr.left;
            }

        }else{
            if(curr.left != null){
                maxXor += Math.pow(2, i);
                curr = curr.left;
            }else{
                curr = curr.right;
            }
        }
    }
    return maxXor;
}

var findMaximumXOR = function(nums) {
    let tr = new Trie();
    for(let i = 0; i < nums.length; i++){
        tr.insert(nums[i]);
    }

    let maxResult = 0;
    for(let i = 0; i < nums.length; i++){
        let temp = findMaxXor(tr, nums[i]);

        maxResult = Math.max(maxResult, temp);
    }

    return maxResult;
};

File: 0421-maximum-xor-of-two-numbers-in-an-array/0421-maximum-xor-of-two-numbers-in-an-array.java
class Solution {
    public static class trieNode{
        trieNode left;
        trieNode right;
    }

    public static void insert(trieNode root, int num){
        trieNode curr = root;

        for(int i = 31; i >= 0; i--){
            int ith_bit = (num >> i) & 1;
            
            if(ith_bit == 0){ // left
                if(curr.left == null){
                    curr.left = new trieNode();
                }
                curr = curr.left;
            }else{
                if(curr.right == null){
                    curr.right = new trieNode();
                }
                curr = curr.right;
            }
        }
    }

    public static int findMaxXor(trieNode root, int num){
        int maxXor = 0;
        trieNode curr = root;

        for(int i = 31; i >= 0; i--){
            int ith_bit = (num >> i) & 1;

            if(ith_bit == 1){ // 1 means we need 0 for it go left

                if(curr.left != null){ // 1 ^ 0 = 1
                    maxXor += Math.pow(2, i);
                    curr = curr.left;
                }else{ // 1 ^ 1 = 0 // bcoz we dont have left
                    curr = curr.right;
                }
            }else{ // 0 means we need 1 for it go right

                if(curr.right != null){
                    maxXor += Math.pow(2, i);
                    curr = curr.right;
                }else{
                    // maxXor = math.pow(2, i) * 0 = 0 because left means 0;
                    curr = curr.left;
                }
            }
        }

        return maxXor;
    }

    public int findMaximumXOR(int[] nums) {
        trieNode root = new trieNode();

        for(int num : nums){
            insert(root, num);
        }

        int maxResult = 0;

        for(int i = 0; i < nums.length; i++){
            int temp = findMaxXor(root, nums[i]);
            
            maxResult = Math.max(maxResult, temp);
        }

        return maxResult;
    }
}